webhook/resources/volume/validator.go

```diff
@@ -156,6 +164,55 @@ func (v *volumeValidator) Update(request *admission.Request, oldObj runtime.Obje
 	return nil
 }
 
+func (v *volumeValidator) validateExpansionSize(oldVolume *longhorn.Volume, newVolume *longhorn.Volume) error {
+	oldSize := oldVolume.Spec.Size
+	newSize := newVolume.Spec.Size
+	if newSize == oldSize {
+		return nil
+	}
+	if newSize < oldSize && !newVolume.Status.ExpansionRequired {
+		return fmt.Errorf("shrinking volume size from %v/%v to %v/%v is not supported", oldVolume.Name, oldSize, newVolume.Name, newSize)
+	}
+
+	newKubernetesStatus := &newVolume.Status.KubernetesStatus
+	namespace := newKubernetesStatus.Namespace
+	pvcName := newKubernetesStatus.PVCName
+	if pvcName == "" || newKubernetesStatus.LastPVCRefAt != "" {
+		return nil
+	}
+
+	pvc, err := v.ds.GetPersistentVolumeClaim(namespace, pvcName)
+	if err != nil {
+		return err
+	}
+
+	pvcSCName := *pvc.Spec.StorageClassName
+	pvcStorageClass, err := v.ds.GetStorageClassRO(pvcSCName)
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return errors.Wrapf(err, "cannot find storage class %v", pvcSCName)
+		}
+		return err
+	}
+	if !*pvcStorageClass.AllowVolumeExpansion {
+		return fmt.Errorf("storage class does not allow the volume expansion")
+	}
+
+	pvcSpecValue, ok := pvc.Spec.Resources.Requests[corev1.ResourceStorage]
+	if !ok {
+		return fmt.Errorf("cannot get request storage")
+	}
+
+	requestedSize := resource.MustParse(strconv.FormatInt(newSize, 10))
+	if pvcSpecValue.Cmp(requestedSize) < 0 {
+		return fmt.Errorf("waiting for pvc %v expansion from %v to %v", pvcName, pvcSpecValue.Value(), requestedSize.Value())
+	}
+	if pvcSpecValue.Cmp(requestedSize) > 0 && !newVolume.Status.ExpansionRequired {
+		return fmt.Errorf("shrinking pvc %v size from %v to %v is not allowed", pvcName, pvcSpecValue.Value(), requestedSize.Value())
+	}
```

Comment:
I am thinking if this is a little bit risky:
Somehow users expand the PVC spec size while the volume is handling other requests and it cannot start expansion immediately. Then the volume probably cannot continue handling anything since the update for other requests will be failed by this check.

BTW, IIRC, Kubernetes does not support shrink PVC spec size.

Reply:
Yes, Kubernetes does not support shrink PVC spec size.
Removed this check statement. Thanks